feat(common): add injection token for default date pipe timezone

crisbeto opened the pull request:
Adds a new injection token called `DATE_PIPE_DEFAULT_TIMEZONE` that allows for the default timezone for all `DatePipe` instances to be specified.

Fixes #43031.

petebacondarwin commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -6,17 +6,20 @@
 import {formatDate} from '../i18n/format_date';
 import {invalidPipeArgumentError} from './invalid_pipe_argument_error';
 
+/** Default timezone to use for all instances of `DatePipe` (such as '+0430'). */
```
I think we should have a bit more documentation here, explaining that this is optional and what happens if you don't provide it. Perhaps giving some examples.

petebacondarwin replied:
Currently looks like https://pr43611-54afbf0.ngbuilds.io/api/common/DATE_PIPE_DEFAULT_TIMEZONE

petebacondarwin commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -166,7 +169,9 @@ import {invalidPipeArgumentError} from './invalid_pipe_argument_error';
-  constructor(@Inject(LOCALE_ID) private locale: string) {}
+  constructor(
+      @Inject(LOCALE_ID) private locale: string,
+      @Inject(DATE_PIPE_DEFAULT_TIMEZONE) @Optional() private defaultTimezone?: string) {}
```
If the token is not provided then I think the DI will inject `null` not `undefined`, right?
So perhaps the signature should be :defaultTimezone?: string|null`.

crisbeto replied:
You're right, but I remember VE having some issues with union types in constructors.

petebacondarwin replied:
Does using a union actually break out tests in VE? If not then I would prefer to be explicit since `null` is actually a likely value to be passed in here.

crisbeto replied:
I didn't try actual tests, but in the past we tried applying `| null` to constructor parameters on Material and VE threw an error.

alxhub replied:
It _should_ be okay as long as you have `@Inject`.

crisbeto replied:
Seems like it handles the `| null` fine so I've switched it. I still had to keep the parameter optional, because there are some usages in g3 that construct the pipe manually.

petebacondarwin reviewed (changes requested):
This is great! I love the use of `@Optional` to avoid any breaking change for direct users of the `DataPipe`.

mary-poppins commented:
You can preview 77e82a8 at https://pr43611-77e82a8.ngbuilds.io/.
You can preview 54afbf0 at https://pr43611-54afbf0.ngbuilds.io/.

petebacondarwin commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -188,8 +193,8 @@ export class DatePipe implements PipeTransform {
-      value: Date|string|number|null|undefined, format = 'mediumDate', timezone?: string,
-      locale?: string): string|null {
+      value: Date|string|number|null|undefined, format = 'mediumDate',
+      timezone = this.defaultTimezone, locale?: string): string|null {
```
For some reason we are not picking up the descriptions of the parameters above in our API docs on AIO...

<img width="693" alt="Screenshot 2021-09-27 at 20 46 01" src="https://user-images.githubusercontent.com/15655/134974996-6a7bc6b3-b0ca-42de-9e32-abaee8f42c1e.png">

But once that is fixed, I think it would be good to update them to indicate that the default is actually injected via the new token.

crisbeto replied:
I'll remove the default value from the function for now and apply the default inline.

crisbeto commented:
The feedback has been addressed.

mary-poppins commented:
You can preview 9d17447 at https://pr43611-9d17447.ngbuilds.io/.

petebacondarwin reviewed (approved):
Ideally we should fix the type of the injected param, but if not possible then LGTM.

crisbeto commented:
@alxhub the feedback has been addressed.

dylhunn commented:
Presubmit is passing: http://test/OCL:401278522:BASE:401291859:1633543923861:f2e6d38f

crisbeto committed:
feat(common): add injection token for default date pipe timezone

Adds a new injection token called `DATE_PIPE_DEFAULT_TIMEZONE` that allows for the default timezone for all `DatePipe` instances to be specified.

Fixes #43031.

crisbeto commented on packages/common/src/pipes/date_pipe.ts:
```diff
@@ -6,17 +6,23 @@
  *
  * Formats a date value according to locale rules.
- * 
+ *
```
Done.

alxhub reviewed (approved):
Reviewed-for: public-api

dylhunn commented:
This doesn't need another presubmit run, since all changes since the last are docs-only

petebacondarwin reviewed (approved):
Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api
